edgelet/edgelet-image-cleanup/src/image_gc.rs

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use std::{collections::HashSet, time::Duration};
+
+use chrono::Timelike;
+use edgelet_core::{ModuleRegistry, ModuleRuntime};
+use edgelet_docker::ImagePruneData;
+use edgelet_settings::base::image::ImagePruneSettings;
+
+use crate::error::ImageCleanupError;
+
+const TOTAL_MINS_IN_DAY: u32 = 1440;
+
+pub async fn image_garbage_collect(
+    edge_agent_bootstrap: String,
+    settings: ImagePruneSettings,
+    runtime: &edgelet_docker::DockerModuleRuntime<http_common::Connector>,
+    image_use_data: ImagePruneData,
+) -> Result<(), ImageCleanupError> {
+    log::info!("Starting image garbage collection task...");
+
+    if !settings.is_enabled() {
+        return std::future::pending().await;
+    }
+
+    let cleanup_time = &mut settings.cleanup_time();
+
+    let diff_in_secs: u32 = get_sleep_time_mins(cleanup_time) * 60;
+    tokio::time::sleep(Duration::from_secs(diff_in_secs.into())).await;
+
+    let mut bootstrap_image_id_option = None;
+    let mut is_bootstrap_image_deleted: bool = false;
+
+    loop {
+        // Try to get the bootstrap image id if we failed on the last(/all previous) run(s)
+        if bootstrap_image_id_option.is_none() {
+            // bootstrap edge agent image should never be deleted
+            if let Ok((id_option, is_image_deleted)) =
+                get_bootstrap_image_id(runtime, edge_agent_bootstrap.clone()).await
+            {
+                is_bootstrap_image_deleted = is_image_deleted;
+                if !is_image_deleted {
+                    bootstrap_image_id_option = id_option.clone();
+                }
+            } else {
+                log::error!("Could not get bootstrap image id");
+            }
+        }
+
+        if bootstrap_image_id_option.is_some()
+            || (bootstrap_image_id_option.is_none() && is_bootstrap_image_deleted)
+        {
+            remove_unused_images(
+                runtime,
+                image_use_data.clone(),
+                bootstrap_image_id_option.clone(),
+            )
+            .await?;
+        }
+
+        // sleep till it's time to wake up based on recurrence (and on current time post-last-execution to avoid time drift)
+        let recurrence = settings.cleanup_recurrence();
+        let delay = recurrence
+            - Duration::from_secs(
+                ((TOTAL_MINS_IN_DAY - get_sleep_time_mins(cleanup_time)) * 60).into(),
+            );
+        tokio::time::sleep(delay).await;
```

Comment:
To run periodic tasks, you want tokio::time::interval instead of calculating and calling sleep yourself. See the watchdog code.

Also instead of running another task to do cleanups, maybe just have the watchdog do it. It's already a periodically-running task.

Reply:
Varun explicitly mentioned that we don't want this to be part of the watchdog when we started talking about this feature (and I agree).

I did look at tokio::time::interval, and I'm blanking on why I had found that it doesn't work.

It might have been because customers want to run this at a specific time, and not just periodically.

(But I'll take another look to see if I missed something, which is definitely possible)

Reply:
I looked at this again, and it _is_ the fact that we have to run the task at a specific time.

But it's also a way of dealing with system restarts -- if there's a restart 5 minutes before when the task is supposed to run, calling sleep yourself allows you to satisfy the ask that the task is supposed to run around a certain time.